### DISCUSSION TITLE
Proposal: Codegen: Return simple objects

### DISCUSSION THREAD

**Discussion by helderco** [dagger/dagger#12508] [Ideas]
## Summary

For simple object types where all their fields are non-parametrized scalars, make it more convenient to access the object’s properties in SDKs by returning the whole object at once as a leaf value.

## Motivation

Our chainable API is pretty simple, but it also feels limiting when you want multiple related fields, especially if they’re a part of a simple object like `EnvVariable`, `Port` or `Label`.

For example, if I have an `EnvVariable` object, and I want both the `name` and `value`, I need to make two extra requests:

```go
envVars, err := ctr.EnvVariables(ctx)
if err != nil {
    return nil
}
for _, env := range envVars {
    name, err = env.Name(ctx)
    if err != nil {
        return nil
    }
    value, err = env.Value(ctx)
    if err != nil {
        return nil
    }
    fmt.Printf("%s=%s\n", name, value)
}
```

Same thing with `Port`, `Label` and others like these.

This could be much simpler by making a **single request** that includes all fields in the result:

```go
envVars, err := ctr.EnvVariables(ctx)
if err != nil {
    return nil
}
for _, env := range envVars {
    fmt.Printf("%s=%s\n", env.Name, env.Value)
}
```
 
While these are just two fields, `Port` has three and there’s an issue for adding a `FileInfo` type (similar to [`fs.FileInfo`](https://github.com/dagger/dagger/issues/3457#issuecomment-1397574645)) that’ll double that (4 - 6 fields):
- https://github.com/dagger/dagger/issues/3457#issuecomment-1397574645
- https://github.com/dagger/dagger/issues/3764

### Definition

This proposal defines a “simple object” as:
- A GraphQL **object type**;
- Where **all fields are leafs** that **don’t have any arguments**;
- Doesn’t have an `id` field (rules out `CacheVolume`, `Socket` and `Secret`).

### Codegen

When a simple object is detected, print it the same as an input type (except it should only be used in outputs):
```diff
// A simple key value object that represents an environment variable.
type EnvVariable struct {
-	q *querybuilder.Selection
-	c graphql.Client
-
-	name  *string
+	// The environment variable name.
+	Name  string
+
-	value *string
+	// The environment variable value.
+	Value *string
}
-
-// The environment variable name.
-func (r *EnvVariable) Name(ctx context.Context) (string, error) {
-	if r.name != nil {
-		return *r.name, nil
-	}
-	q := r.q.Select("name")
-
-	var response string
-
-	q = q.Bind(&response)
-	return response, q.Execute(ctx, r.c)
-}
-
-// The environment variable value.
-func (r *EnvVariable) Value(ctx context.Context) (string, error) {
-	if r.value != nil {
-		return *r.value, nil
-	}
-	q := r.q.Select("value")
-
-	var response string
-
-	q = q.Bind(&response)
-	return response, q.Execute(ctx, r.c)
-}
```

Then, if a field returns a simple object, select all fields and use it directly in responses:

```diff
 // Retrieves the list of environment variables passed to commands.
 func (r *Container) EnvVariables(ctx context.Context) ([]EnvVariable, error) {
        q := r.q.Select("envVariables")

        q = q.Select("name value")

-       type envVariables struct {
-               Name  string
-               Value string
-       }
-
-       convert := func(fields []envVariables) []EnvVariable {
-               out := []EnvVariable{}
-
-               for i := range fields {
-                       out = append(out, EnvVariable{name: &fields[i].Name, value: &fields[i].Value})
-               }
-
-               return out
-       }
-       var response []envVariables
+       var response []EnvVariable

        q = q.Bind(&response)
-
-       err := q.Execute(ctx, r.c)
-       if err != nil {
-               return nil, err
-       }
-
-       return convert(response), nil
+       return response, q.Execute(ctx, r.c)
 }
```

We’re just moving that implementation detail to the outside, removing the need for passing context and capturing error.

### Detection

With https://github.com/dagger/dagger/issues/5226, **all SDKs should be able to reuse** the same detection logic during introspection. But for now, I wrote a simple introspection [command](https://github.com/dagger/dagger/pull/5300) in Python to test the theory and as of now, these are the types that match:

```graphql
"""A simple key value object that represents an environment variable."""
type EnvVariable {
  """The environment variable name."""
  name: String!

  """The environment variable value."""
  value: String!
}

"""A flag accepted by a project command."""
type ProjectCommandFlag {
  """Documentation for what this flag sets."""
  description: String

  """The name of the flag."""
  name: String!
}

"""A simple key value object that represents a label."""
type Label {
  """The label name."""
  name: String!

  """The label value."""
  value: String!
}

"""A port exposed by a container."""
type Port {
  """The port description."""
  description: String

  """The port number."""
  port: Int!

  """The transport layer network protocol."""
  protocol: NetworkProtocol!
}
```

<details><summary>See <strong>code</strong> that produced this 👆</summary>

```python
from graphql import GraphQLNamedType, get_named_type
from graphql.utilities.print_schema import print_filtered_schema

import dagger
from dagger.codegen import (
    is_object_type,
    is_output_leaf_type,
)
from dagger.server import command


@command
def find(client: dagger.Client) -> str:
    """Find all simple objects in the schema."""
    schema = client._ctx.schema._schema
    out = print_filtered_schema(schema, lambda _: False, is_simple_object)
    print(out)
    return out


def is_simple_object(t: GraphQLNamedType) -> bool:
    t = get_named_type(t)
    if t.name.startswith("_") or not is_object_type(t):
        return False
    return not any(
        f
        for f_name, f in t.fields.items()
        if f_name == "id" or f.args or not is_output_leaf_type(f.type)
    )
```

</details>

## Backwards compatibility

This is an obvious breaking change **in SDKs** (not in the API) because the signature for handling the above types is different. However, this is **only a problem in Go** but not in Python and Node.js because it doesn’t even work today (see more about this in [Rationale: Handling lists of objects](#handling-lists-of-objects)).

While we could introduce half measures to keep some sort of compatibility I think it’s better to just **cluster this with other breaking changes** we need to make in a single release.

### <a name=”handling-lists-of-objects”></a>Handling lists of objects

There’s a long standing issue about SDKs not being able to return lists of objects:
- https://github.com/dagger/dagger/issues/3457

Returning lists of objects were a known limitation in our chainable objects in the SDKs, so we had an undocumented convention for not supporting that until we had a good solution, but eventually it got into the API without having tests via the SDKs.

So to fix the introduced bug, @sipsma suggested:

> If we are willing to trade more complexity for having the ability to return objects in lists maybe we could figure something out around autoselecting all trivial scalars for those cases, though I don't know if we'll have to apply that fix on a case-by-case basis (ugly) or if there's a way of cleanly generalizing it. Vague idea, might be a deadend, but in general it would be nice to support these return types.

\- Reposted by @aluzzardi in https://github.com/dagger/dagger/issues/3457#issuecomment-1316112171

That idea has been adopted as the fix in Go:
- https://github.com/dagger/dagger/pull/4683

However, the implementation isn’t ideal. When a list is returned, the fields are internally pre-selected and filled on private struct fields. When they are accessed, if the cached value exists then it’s short circuited:

```go
type EnvVariable struct {
	// ...
	name  *string
	value *string
}

func (r *EnvVariable) Name(ctx context.Context) (string, error) {
	if r.name != nil {
		return *r.name, nil
	}
	// ...
}

func (r *EnvVariable) Value(ctx context.Context) (string, error) {
	if r.value != nil {
		return *r.value, nil
	}
	// ...
}
```

That means there’s useless verbosity when accessing these fields, i.e., context isn’t actually needed and the `err` is always `nil`:

```go
name, err = env.Name(ctx)
if err != nil {
    return nil
}
value, err = env.Value(ctx)
if err != nil {
    return nil
}
```

You can ignore the errors in this case, but it’s an implementation detail and feels unsafe to depend on:

```go
name, _ = env.Name(ctx)
value, _ = env.Value(ctx)
```

> Yeah, I would have preferred to directly access the property instead of passing a context and unpacking the error but it's unfortunately not possible because of the current SDK API design :/

\- Originally posted in https://github.com/dagger/dagger/pull/4683#issuecomment-1452262848

Meaning, this was desired instead:

```go
name = env.Name
value = env.Value
```

I haven’t fixed this in Python yet because I wanted a better solution[^1], that’s why I came up with this proposal, which also solves some use cases for https://github.com/dagger/dagger/issues/3577.

[^1]: I no longer intend to wait on a solution to fix it in Python, just haven’t made it a priority yet but will soon.

Although it’s not that verbose in Python and Node.js:

```python
name = await env.name()
value = await env.value()

# vs

name = env.name
value = env.value
```

But this proposal’s design would also works for any field that returns one of these objects directly, **not just on lists of objects**, which addresses the following concern:

> Also creates a disparity between `Object` (need to call `Object.Field(ctx)` -- it's lazy) and `[Object!]!` (need to call `Object.Field` -- it's not lazy). We could have both APIs be the same, but the behavior would be different (lazy vs non-lazy) which can be confusing. Implementation would also need to be different (return "pre-fetched" field vs query the API).

\- Originally posted by @aluzzardi in https://github.com/dagger/dagger/issues/3457#issuecomment-1316112171

I’m proposing that when these objects are returned, they’re always handled like a leaf, not just on lists.

#### SDK helper

At first I considered an SDK helper to hide the verbosity and boilerplate of making [concurrent requests](https://github.com/dagger/dagger/issues/4766) for the multiple sibling fields, but my pain point at the time was for getting exec related fields (i.e., `exitCode`, `stdout`, `stderr`). That’s why I added an `ExecResult` example to the first version of this proposal, but there’s been other solutions in the meantime.

#### Explicit declaration with API directive

GraphQL has *input* object types, used only as arguments to fields:

```graphql
input BuildArg {
  name: String!
  value: String!
}
```

Codegen is already converting those into simple structs/objects:

```go
type BuildArg struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}
```

I would like to have an `output EnvVariable {}` type to do the same but for the return values. Since that doesn’t exist, we can simulate it with a directive:

```graphql
directive @output on OBJECT

type EnvVariable @output {
  name: String!
  value: String!
}
```

Which would produce a similar struct/object to input types:

```go
type EnvVariable struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}
```

This solution makes detection much simpler in codegen and explicit in the API.

It’s not my first choice because it’s an added burden on API authors to choose correctly, potentially leaving some **inconsistencies** which would need extra validation in the API itself (e.g., using on a type that has non-leaf fields or fields with arguments, forgetting to use it in a new type and introducing a break in the convention….). But also when used in lists there’s an implicitness there already where this is actually required.

I chose to go with the implicit version because it’s better at enforcing the rules. However, this can become an option to have better control if in the future we find the need for it for any reason. By then, maybe other kinds of directives would be better, like marking a field as always being prefetched and used as a property, or marking a field as expensive so it gets ruled out as a property.

By then it won’t be a breaking change.

#### Alternative value object type

I thought about adding a second or alternative for these object types in the SDK clients, but that adds unnecessary fragmentation or inconsistency:

```go
type EnvVariableValue struct {
	Name  string
	Value string
}

type EnvVariable struct {
	// ...
	name  *string
	value *string
}

func (r *EnvVariable) Name(ctx context.Context) (string, error) {
	// ...
}

func (r *EnvVariable) Value(ctx context.Context) (string, error) {
	// ...
}
```

## Drawbacks

- It’s a breaking change, but we can cluster it with other breaking changes.
- Small benefit? Today we only have 4 object types with 2-3 fields but that’ll change (e.g., [`FileInfo`](https://github.com/dagger/dagger/issues/3457#issuecomment-1397574645)).
- We can find later on that we need more control over the selection (e.g., what if a leaf isn’t a “trivial” scalar?), but we can add directives if it comes to that.
- Doesn’t entirely fix [selecting sibling fields](https://github.com/dagger/dagger/issues/3577) where a proper solution would fix this as well.
- Doesn’t entirely fix [returning lists of objects](https://github.com/dagger/dagger/issues/3457) if you need to list an object that has at least one field that’s not a scalar.
- Can be confusing that some fields are properties (`Object.Field`) while others are functions (`Object.Field(ctx)`)

## Open question

There’s only one (very recent) case for the last point above in `ProjectCommand`, unless we include the `id` field (as a method) and also recursively resolve all the fields that list simple objects themselves. This isn’t hard to do, it’ll actually still make the implementation simpler compared to today. Would this be worth doing?

https://github.com/dagger/dagger/blob/3bb89e9ff1478077d271fa198d54edc96437487b/core/schema/project.graphqls#L30-L46

**Comment by helderco** [comment]
I just updated the proposal which was long overdue.

I removed the example on `ExecResult` which was the main reason I wanted this in the first place, but there's other solutions for that and I no longer think it worth it to have a wrapping type for `exitCode`, `stdout` and `stderr` especially if the output could be huge and you don't want all those fields when you just need one.

**Comment by gerhard** [comment]
Having spoken to @helderco sync, we are thinking of not including this in the upcoming breaking changes release since more discussions are needed. FTR:
- https://github.com/dagger/dagger/discussions/5374

**Comment by TomChv** [comment]
@helderco This issue can help to improve #5557 integration (See https://github.com/dagger/dagger/pull/5557#issuecomment-1687440697)

Should we start working on It? I understood your proposal and have an implementation in mind for Node and Go.
This would make our generators a bit more complex and specific (since we have special function for what you call simple object).

**Comment by helderco** [comment]
> Should we start working on It?

We should agree on design first.

**Comment by TomChv** [comment]
I kinda like the design proposed in your specification! I don't know if you have something different in mind for now?
We can have a quick sync chat if you're free to agree on the design :)

**Comment by helderco** [comment]
I meant we should get more feedback from the team if it's worthwhile considering the drawbacks and open question.

**Comment by TomChv** [comment]
@vito @sipsma @shykes @aluzzardi Waiting for your feedbacks then :D

**Comment by vito** [comment]
Sorry for the slow feedback on this! I think it's important; I've been thinking about related issues a lot lately, hoping something forces the issue, and it seems like this does, so yay!

I like your proposed solution the most, but there's one part that I'm hesitant about:

> * Doesn’t have an id field (rules out CacheVolume, Socket and Secret).

I'm hesitant here because as part of the Zenith spike I gave an ID to what is otherwise a simple object: [`EnvironmentCheckResult`](https://github.com/sipsma/dagger/blob/zenith/core/schema/environment.graphqls#L346). It's ultimately just so that it can be serialized/deserialized in a typed way (Zenith also changes IDs to `core.DirectoryID:eyJ...`).

I'm not sure if we'll stick with that approach, but the fact that I had the option to once I made it ID-able makes me consider going full GraphQL purist and saying everything is a [Node] (has an `id`) and there is no such thing as a "simple data object." Another way of thinking about the above use case is that the Environment produced a DAG `Node` and needs to return its global database ID to the caller, which then loads the ID back into the `Node`.

[Node]: https://graphql.org/learn/global-object-identification/#node-interface

At that point we could also unify on a single `ID` type, since the ID _contains_ the type. We would _probably_ want to do this, since "everything has an ID" + "every ID has its own type" would mean a ton more ID types.

But I still don't want to have to call methods just to get simple scalar static fields from a list of objects, so yeah, this is still important!

Could we do something like the `@output` directive but declared on individual fields rather than an entire type?

```graphql
directive @field on FIELD_DEFINITION

type EnvironmentCheckResult {
  id: EnvironmentCheckResultID!

  name: String! @field # or @prefetch / @immutable / @static / @simple / 🤷‍♂️
  withName(name: String!): EnvironmentCheckResult!

  success: Boolean! @field
  withSuccess(success: Boolean!): EnvironmentCheckResult!

  output: String! @field
  withOutput(output: String!): EnvironmentCheckResult!

  subresults: [EnvironmentCheckResult!] @field
  withSubresult(result: EnvironmentCheckResultID!): EnvironmentCheckResult!
}
```

Each SDK would then know to pre-fetch `name`, `success`, and `output` so they can be exposed on the struct:

```go
type EnvironmentCheckResult {
  Name string
  Success bool
  Output string
}

func (res *EnvironmentCheckResult) ID(ctx context.Context) (EnvironmentCheckResultID, error) {
  // ...
}
```

I remember "ID-able vs. non-ID-able" being something we put a _lot_ of leverage on, so this might be a game of jenga, but it seems worth exploring since it more closely aligns us with GraphQL.

It seems like there are a lot of benefits to "everything is a `Node`":

* It's always easy to pass around _anything_ in the DAG.
  * Even "simple objects" can just literally put all the info in the ID, like above.
  * In the future we could move things to a database if necessary. We've always talked about wanting to shrink these IDs.
* We can add non-static methods to any type whenever we want to.
  * This is easy for methods that are wholly derived from the other fields.
  * Otherwise you might need to actually be using a database of some kind, but you can at least transition the ID to that when the time comes.
* I should sleep, but there are probably more. 😅
* edit: We don't need to distinguish "input" vs "output" types in GraphQL - just leverage our existing SDK syntax sugar for ID-able objects. (Any returned object can be passed as an input by ID.)

> Sidebar: I've also been wondering what the ID format should look like and what guarantees it should hold. We've been making it up as we go along. [RFC-6920](https://posts.summerti.me/rfc-6920/) is interesting. Another idea has been to indicate whether the ID refers to a "pure" value (contains no non-reproducible values like local dirs, secrets, etc). We would also want to decide if they are publicly shareable or if they are only valid within a session. Lots of overlap with #3923 too.

**Comment by gerhard** [comment]
We just talked about this with @helderco & are thinking that we will come back to this Q1 2024. Leaving open for now. Needs further discussion.

**Comment by helderco** [comment]
@vito, I meant to reply before going on vacation in September but then forgot 😅. I'll reply soon.

**Comment by helderco** [comment]
Now that we have modules, with a distinction between *fields* and *functions* in an object, I’m much more inclined to the explicit `@field` directive approach. In the CLI we want to be able to print the fields of an object instead of returning an error when the chain ends in one:
- https://github.com/dagger/dagger/issues/6515

That, however, is only available to user modules because *core* doesn’t have this distinction in its object types. GraphQL objects only have fields. Today we'd need to hardcode special handling.

To support [core chaining](https://github.com/dagger/dagger/issues/6078) in the CLI, core types are being *TypeDef*’ed, also in preparation of making *core* a module in the future. At the moment, all core objects have only functions. With `@field` we would be declaring which ones can be a *module field*. I should probably put this into a separate issue, more focused on the *module representation* aspect, and keep this one focused on the SDK DX of returning an object with pre-populated fields. These two concepts overlap, but I haven’t spent enough time checking if they’re compatible together.

Simple rules/validation for choosing where to put a `@field`:
- Doesn’t have any arguments;
- Returns a “trivial” value;
- Think “state”;

Returning a `string` scalar isn’t enough. It has to be something that’s not expensive to return; however, this needs to be discerned by the maintainer. For example, `Container.stdout` returns a `string`, but it triggers execution, which can take a long time. It’s excluded anyway since it takes arguments.

Anything in `Container` that returns something from `ImageConfig` is “static” though, so it’s easy to fetch (user, workdir, env vars, exposed ports, labels, default command…). For `File` you have `name` and `size`. For these kinds of objects, “trivial” means metadata. For a simple object like `EnvVariable` it’s the actual data.

**Comment by shykes** [comment]
We don't want a distinction between field and function. It doesn't exist in the underlying graphql spec, and it would be unnecessary cognitive load for developers to try to understand the difference.

**Comment by helderco** [comment]
This distinction wouldn't show in the DX, not as fields anyway. It's just a way to mark some graphql fields as "trivial" and allow some optimizations, particularly pre-selecting in certain cases like on lists or when printing a core object in the CLI.

**Comment by vito** [comment]
Don't we already have a distinction in the DX for modules in particular? In the Go SDK for example you have fields on the struct for an object, while functions are methods on the struct. How would you manage state without a fields vs. functions distinction?

**Comment by helderco** [comment]
Indeed. When we turn core into typedefs, they're all functions, no declared "state". When a module with fields extends the schema, they become functions without arguments. We could preserve this information with the directive.

**Comment by vito** [comment]
For continuity from Discord conversations with @shykes:

We agree that there's a need for state and that there is effectively a fields vs. functions distinction internally, but @shykes 's point is that fields should just be an implementation detail of modules, not distinguished in any way through the API. The main take-away is that it should be easy for a module author to switch something from a field to a function internally without any impact on users.

So I think to address this we'll need better client support for selecting multiple fields. Right now you have to drop down to a handwritten query string. We could explore a totally different style of client based on a query building approach - there's consensus in curiosity, but not commitment. Or maybe there's a middle ground where we can integrate multiple field selection into the existing SDK in a natural way.

**Comment by helderco** [comment]
> We could explore a totally different style of client based on a query building approach

I summarized that in https://github.com/dagger/dagger/issues/4668#multiple. By that I mean separating query building from execution, as in:
```go
// build query, doesn't execute
q := dag.Container().Build(src).Stdout()

// only one way to execute query!
err := dagger.Query(q)
```

From my POV, the big issue with that approach was we didn't find a good way to extract multiple values, from a multiple selection. Solomon had a few ideas but they didn't cross language very well. We were a bit stuck discussing this (see https://github.com/dagger/dagger/issues/3555#issuecomment-1301327344), so it was decided at the time to just move on with the more simple solution, which is what we have today.

After a while, when people were expressing confusion over the implicitness of execution on leafs, contrary to just building a query, we were thinking that the query builder approach could make it easier to understand for people, but also that it was too big a change and it was too late.

> Or maybe there's a middle ground where we can integrate multiple field selection into the existing SDK in a natural way.

Yeah, in the meantime, this often pops up in my head, i.e., selecting multiple siblings with the current DX, but the thought is like a doctor's visit, which doesn't stay long. 😄

I usually gravitate towards something to the tune of generating a second object type for each object type, that only includes the leafs but still chainable to allow for multiple selection:
```go
ctr, err := dag.Container().
    From("alpine").
    WithExec([]string{"printenv"}).
    Leafs().  // bikeshedding non-withdstanding
    Stdout().
    Stderr().
    Sync(ctx)  // or anything else to Execute
```
Essentially there would be a `Container.Leafs -> &ContainerLeafs{}` where `ContainerLeafs.Stdout()` just adds the sibling selection to the query builder and returns "self", like the other non-leafs.

At this point, when I need to figure out the best way to get the multiple values in a way that is type safe and has a good DX over multiple languages, my 🧠 jumps to another branch in the 🐒 🌴. 🙂
